big_tests/tests/carboncopy_SUITE.erl

```diff
@@ -147,8 +146,8 @@ dropped_client_doesnt_create_duplicate_carbons(Config) ->
       fun(Alice1, Alice2, Bob) ->
               Msg = escalus_stanza:chat_to(Bob, <<"And pious action">>),
               given_client_logs_out(Config, Alice2),
-              _Pres = escalus_client:wait_for_stanza(Alice1),
-
+              escalus:assert(is_presence_with_type, [<<"unavailable">>],
```

Comment:
@zofpolkowska unfortunately that might not be enough. Apparently, we had situations on Travis when user got "unavailable" but did not really get deleted on the server side.

Reply:
so do we have to use `timer:sleep` here for now?

Reply:
Ok, after diving into the code I claim that the solution you provide is OK. It is because carboncopy does not rely on the session table but rather keeps its internal list of users to broadcast to. And this list updated on hook `unset_presence_hook` which is run just before broadcasting presence "unavailable" to other users/resources. That makes the solution valid. The problem is that if we change the way carboncopy updates this list (or rather the moment it does it) we won't catch a possible race condition here, which is updating the list AFTER we broadcast presence. I believe we should write a regression test for such a scenario. Of course that should not be in the scope of this PR.

Reply:
@erszcz it would be nice if you also glanced at the problem and correct me if I'm wrong.

Reply:
My interpretation is a bit different.

Let's start with [`escalus_client:stop/2`](https://github.com/esl/escalus/blob/e7eece237a56560add06127bc9ed47d423e88dcc/src/escalus_client.erl#L74-L77) - it calls [`escalus_connection:stop/1`](https://github.com/esl/escalus/blob/e7eece237a56560add06127bc9ed47d423e88dcc/src/escalus_connection.erl#L253-L261), which properly ends the XMPP stream with regard to the connection type (abstraction over `Mod`, which might be `escalus_tcp`, `escalus_bosh`, or `escalus_ws`), but does not send `<presence type="unavailable" .../>` before ending the stream. Once XMPP stream is terminated, `escalus_connection:stop/1` also calls `Mod:stop/1` which ends the connection(s) itself.

On the server side, `mod_carboncopy` deregisters a user on the `unset_presence_hook`, but I don't agree with:

> carboncopy does not rely on the session table but rather keeps its internal list of users to broadcast to

On the contrary, it relies on the session table, which is accessed in [`mod_carboncopy:remove_connection/5`](https://github.com/esl/MongooseIM/blob/c3d2ffe66b297560f2da03cd53cb37c9c7dddc5b/src/mod_carboncopy.erl#L188-L190) fired by the `unset_presence_hook`. This hook is triggered in [`ejabberd_sm:unset_presence/7`](https://github.com/esl/MongooseIM/blob/c3d2ffe66b297560f2da03cd53cb37c9c7dddc5b/src/ejabberd_sm.erl#L351-L353) and [`ejabberd_sm:close_session_unset_presence/6`](https://github.com/esl/MongooseIM/blob/c3d2ffe66b297560f2da03cd53cb37c9c7dddc5b/src/ejabberd_sm.erl#L366-L368).

The interaction of `mod_carboncopy` and `ejabberd_sm:unset_presence/7` is tested in `unavailable_resources_dont_get_carbons/1`, so here we're interested in `close_session_unset_presence/6`.

`ejabberd_sm:close_session_unset_presence/6` is called in two places in `ejabberd_c2s` - both times in `ejabberd_c2s:terminate/3`:

- [when terminating a resource that was replaced by a new connection](https://github.com/esl/MongooseIM/blob/c3d2ffe66b297560f2da03cd53cb37c9c7dddc5b/src/ejabberd_c2s.erl#L1654),
- [in the general case](https://github.com/esl/MongooseIM/blob/c3d2ffe66b297560f2da03cd53cb37c9c7dddc5b/src/ejabberd_c2s.erl#L1691).

In both cases the presence of type `unavailable` is sent after `ejabberd_sm:close_session_unset_presence/6` returns, the `unset_presence_hook` fired in `close_session_unset_presence` call is synchronous, therefore when a remaining user's resource receives the presence it's guaranteed that duplicate carbon copies won't be created.

Reply:
Of course, you’re right, I missed the point we are “brutally” killing the connection and not sending presence beforehand. Sorry for that.